### ISSUE TITLE
Support `restart` in terminated event in debug adapter

### ISSUE THREAD

**Issue by nturinski** [microsoft/debugpy#577]
It would be really handy if Python launch configurations had a similar property to Node.js's restart property. For reference: https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_restarting-debug-sessions-automatically-when-source-is-edited.

The use case is whenever a python file is edited and saved, it causes the debugger to detach so it'd be nice to have an option to automatically re-attach the debugger in those instances.

**Comment by fabioz** [comment]
I'll start to work on this issue.

I'll rephrase what I understood (so that someone can correct me if I got the wrong idea).

The idea is that given an attach launch configuration that has `"restart": true`, if the target process disconnects abruptly (i.e.: because it was killed), the adapter should ask VSCode to try to rerun the same attach launch configuration (by sending a `"terminated"` event with a `"restart": <some_value>`, which will be later received in the new `attach` request as `"__restart"`).

**Comment by karthiknadig** [comment]
@fabioz That is correct.

**Comment by fabioz** [comment]
To give some feedback, I tried just passing the `"restart": "restart"` in the `"terminated"` event, but there are some complications...

The debug adapter in this case is started by the server (python process) as a separate process and starts communicating with VSCode at the specified port (say `5678`) and then creates a separate connection to talk to the server process at a different random port (say `78787`).

When it sends the `"terminated"` event with a `"restart": "restart"`, the debug adapter is kept alive and VSCode keeps communicating with it (it wasn't started by VSCode in the first place, so, that seems fair, VSCode probably only stops what was on his side -- but I was expecting the socket connection to be terminated, which wasn't the case). 

This means that the new connection made fails when trying to start a new adapter to listen in the specified port (`5678`) and a new connection can't be made to the existing adapter because we don't know its port (it's also in a bit of a weird place where it already waited for `servers.wait_until_disconnected()` and is waiting for `sessions.wait_until_ended()`).

I tried sending the `"terminated" with "restart"` and closing the connection/exiting right afterwards but then I couldn't see VSCode connecting back on the newly spawned adapter.

Also, I can see that if the adapter process is not killed, VSCode itself keeps the connection open and issues a new `attach` request to it.

This doesn't work due to 2 things:
1. the new `listen` raises an exception because it'll try to spawn a new adapter and open a new server connection in a port that is already open (`5678`)).
2. even if we didn't create a new adapter, we don't know with which port we should connect to talk to the existing adapter.

So, I need to investigate a bit more on why VSCode is trying to reuse that connection (I thought that VSCode would just disconnect from `5678` -- in which case the adapter should exit and things should just work), but it seems this wasn't the case.

**Comment by fabioz** [comment]
I finished my investigation on this... Fixing this will need changes on both `debugpy` as well as `vscode-python`.

What was happening is that `debugpy` kept the server socket open after the initial connection, and `vscode-python` ended up connecting back to the old adapter (it's expected that it's kept open so that the user can disconnect/reconnect multiple times after the initial attach). Closing the server socket / exiting the adapter on the `debugpy` side does the proper thing.

The problem is that later on, if the connection isn't readily available (which happens because the process that'll listen again takes a bit of time to start-up and get to `debugpy.listen` again), `vscode-python` will not try to reconnect given a timeout.

i.e.: in: https://github.com/microsoft/vscode-python/blob/2021.6.944021595/src/client/debugger/extension/adapter/factory.ts#L50, it uses a `vscode.DebugAdapterServer` and it doesn't seem to have any option to try to reconnect or give any timeout, so, the connection ends up failing because the new connection isn't promptly available.

So, given that I'll proceed to do the changes on `debugpy`, but I'm still unsure of the best way to fix this on the `vscode-python`... ideally there'd be a readily-available way to create a `vscode.DebugAdapterServer` that tries to reconnect if the connection isn't readily available (as a note, I can't see any way to differentiate that the connection in https://github.com/microsoft/vscode-python/blob/2021.6.944021595/src/client/debugger/extension/adapter/factory.ts#L50 is doing a restart or not, so, this would probably also change the regular attach mode).

**Comment by fabioz** [comment]
Note: I'll have leave this on hold for a bit...

Some mental notes for when I (or someone else) checks it:
1. `debugpy.adapter.clients.stop_serving()` must be called before the `terminate` with `restart` is sent to the client so that the old adapter does not respond to the connection.
2. We should add a test with a case with multiprocessing when multiple connections are done (and thus `terminate` with `restart` should only be sent after the last one exits).
3. We shouldn't call `debugpy.adapter.clients.stop_serving()` in the regular case (when the VSCode disconnects), only in the case where all processes exit and the user used an `attach` launch configuration with `restart` as in the regular case the user can connect back to the same socket.
4. `"terminated"` is currently called at `debugpy.adapter.sessions.Session._finalize`, but we need to be careful on the heuristics for which `"restart"` should be passed (it may need more info than it has now).
5. Currently in debug mode in `vscode-python`  it's possible to make it work by adding a breakpoint to https://github.com/microsoft/vscode-python/blob/2021.6.944021595/src/client/debugger/extension/adapter/factory.ts#L50 and letting it run only after a new connection is set up in `debugpy`.

**Comment by int19h** [comment]
I actually wonder if we can make this work without respawning the adapter, rather than trying to hack around the inherent problems with asynchronously connecting to it. So far as I can tell, there are two problems with this in the existing implementation:

0. The adapter doesn't expect a connection to be reused after receiving "disconnect" on it. Some parts of our implicit state machine that handles "initialize", "configurationDone" etc will likely break.

0. When the debuggee is restarted, it'll try to call `debugpy.listen()` again, which always tries to spawn a new adapter on the specified port, and fails if something else is already using that port. There's no way for the debug server to figure out that an existing adapter is already listening on that port, and reconnect to it.

The first problem should be straightforward to handle by resetting all the affected state. The second one is trickier, but I think it's possible for the new server and the original adapter to coordinate. While the new server doesn't know the port on which it should connect to the adapter *as a server*, it does know the port on which the adapter will be listening for clients, since that's passed to `debugpy.listen()` as an argument. If we add a custom client request to retrieve endpoint info, the server could connect to the adapter *as a client* first, and use that custom request to get the other port, and then connect to it normally.

For that matter, separate ports might not even be necessary. We currently use them to distinguish incoming client and server connections, but this can also be done based on the first message received on it - the server will simply send a custom one identifying it as such, and the adapter listener will then create an instance of `Client` or `Server` to process that connection as needed. In this case, the server will always have the necessary info to reconnect.

**Comment by int19h** [comment]
We also need to consider the scenario where there are two *independent* but overlapping debug sessions, each with its own adapter. If the user forgets to specify different port numbers, currently, this will fail and report an error when they try to start the second server. But if we do the above, it will end up quietly attaching to the adapter from the first session, and the whole thing will be treated as one large multiproc session, which would be rather surprising. 

The problem is that the adapter is unable to distinguish between the servers to figure out if the server connecting to it is a new instance that logically corresponds to another session that ended earlier, or it's a completely new one. If the actual restart process provided some facility to pass information from the old server to the new one (like `"__restart"` does for clients), we could use that to maintain some sort of persistent ID. If there's no such facility, I think the ID would have to be provided externally; that is, we'd have to add the ability to pass it to `--listen` / `debugpy.listen()`.

Or perhaps this is best handled by using `--connect` instead? In that scenario, the adapter port to which the server connects isn't random, but is specified by the user. So when the server process restarts with the same arguments, it'd automatically reconnect to the same adapter; and `--listen` wouldn't support restarts. @ejizba, would such a constraint be acceptable for your scenario?

**Comment by snake-py** [comment]
So as someone who just came here - I see you guys are all the way in a discussion about how it could be implemented and seeing the last update is a year ago - I thought I check where the feature request is at. 

I am running Django inside docker and it is an attached process. As an outsider, I do not see why this causes the debuggy process to close. 

Or should this already work and my configs are just wrong?

**Comment by int19h** [comment]
We still don't support "restart", so if your scenario relies on that, it would be expected to not work.

**Comment by fabioz** [comment]
@int19h I was thinking about taking a look into the approach of doing the restart while keeping the debug adapter process live. Is this Ok or do you have plans to work on that already?

**Comment by int19h** [comment]
@fabioz I have some WIP code for this already.

**Comment by int19h** [comment]
Based on the discussion above and in #1212, I'm going to handle this piecemeal. Specifically, the "attach"{"listen"} scenario (i.e. where the debuggee connects to the client) will be first - it's much simpler and with fewer moving bits to synchronize because VSCode is responsible for adapter lifetime in that case; yet it also seemingly covers all the practical scenarios. The complexity induced by the requirements of doing this for  "attach"{"connect"} (i.e. when the debuggee is listening) does not feel like it's worth it, and it's not clear wrt "launch", either.
